### PR TITLE
Complete implementation

### PR DESCRIPTION
**Pull request by jaguirresza**
Add missing endpoints and simplify current implementation.

@mtavano review this please.

### CODE REVIEW

File: khipu.go [hunk 1]
```diff
@@ -1,20 +1,85 @@
...
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const basePath = "https://khipu.com/api/2.0"
```
Comment (mtavano) [inline khipu.go]: What do you think about to use a url.URL instead a raw string??

**Comment by mtavano** [comment]
I feel that you have taken my son, changed his name, changed his clothes and taught him to speak another language 😂   but ☝🏻, it looks really good according sigiriya's conventions


I think that we must to discuss about one of the important fields on Payment struct that I've omitted. According khipus docs, we can set an expiration date to each payment. I've omitted this field because i thought that 3 hours is a reasonable time to a payment application but, if we want to gave our users a longest lapse we need to add this field to our struct.

cc RFC:  @jorgeaguirre @sebastianvera @rodrwan

File: banks.go [hunk 2]
```diff
@@ -2,8 +2,8 @@ package ghipu
 
 import "net/http"
 
-// Bank represents the schema of bank defined by khipu
-type Bank struct {
+// BankItem represents the schema of bank defined by khipu
+type BankItem struct {
```
Comment (mtavano) [inline banks.go:6]: In the first version of Ghipu had declared the structure with this name, however, sebastian recommended me to change it

File: payment.go [hunk 3]
```diff
@@ -86,56 +87,66 @@ func unmarshalJSON(r io.ReadCloser, v interface{}) error {
...
+		"notify_url":  {p.NotifyURL},
+		"contact_url": {p.ContactURL},
+
+		"notify_api_version": {p.NotifyAPIVersion},
+		"integrator_fee":     {p.IntegratorFee},
 	}
 
-	if p.SendEmail {
-		form.Set("send_email", "true")
+	if !p.ExpiresDate.IsZero() {
```
Comment (mtavano) [inline payment.go]: rly like this! 🤘🏻👍🏻

File: khipu.go [hunk 4]
```diff
@@ -105,15 +98,21 @@ func (hc *httpClient) PostForm(path string, values url.Values) (*http.Response,
...
+var percentEncode = strings.NewReplacer(
+	"+", "%20",
+	"*", "%2A",
+	"%7A", "~",
+)
+
 func (hc *httpClient) signRequest(req *http.Request, values url.Values) {
 	var buff bytes.Buffer
 	buff.WriteString(url.QueryEscape(req.Method))
 	buff.WriteByte('&')
```
Comment (mtavano) [inline khipu.go]: why this is not a `.WriteString` function?? 🤔

File: khipu.go [hunk 5]
```diff
@@ -1,20 +1,124 @@
...
+// Client represents an khipu's REST API.
+type Client struct {
+	PaymentService
+	BankService
+}
+
+// NewClient returns an instance of khipu that is the client to make payment request
+func NewClient(secret, receiverID string) *Client {
+	hclient := httpClient{
+		client: &http.Client{},
```
Comment (mtavano) [inline khipu.go:34]: why here you'r not using a http.DefaultClient ??

File: payment.go [hunk 6]
```diff
@@ -1,6 +1,89 @@
...
+	"time"
+)
+
+type PaymentService struct {
+	client *httpClient
+}
+
+func NewPaymentService(secret, receiverID string) *PaymentService {
+	client := httpClient{
+		client: &http.Client{},
```
Comment (mtavano) [inline payment.go]: same here, why not use a simple `http.DefaultClient`instead a default construction of a client?

File: banks.go [hunk 7]
```diff
@@ -1,11 +1,44 @@
...
+	Banks []BankItem
+}
+
+type BankService struct {
+	client *httpClient
+}
+
+func NewBankService(secret, receiverID string) *BankService {
+	client := httpClient{
+		client: &http.Client{},
```
Comment (mtavano) [inline banks.go]: why are you using a default construction instead of `http.DefaultClient`?

File: khipu.go [hunk 8]
```diff
@@ -1,20 +1,123 @@
...
+var percentEncode = strings.NewReplacer(
+	"+", "%20",
+	"*", "%2A",
+	"%7A", "~",
+)
+
+func (hc *httpClient) signRequest(req *http.Request, values url.Values) {
+	var buff bytes.Buffer
+	buff.WriteString(url.QueryEscape(req.Method))
+	buff.WriteByte('&')
```
Comment (mtavano) [inline khipu.go:167]: why are you not using a `.WriteString` function?  is there a special reason to use `.WriteByte` ?

File: khipu.go [hunk 9]
```diff
@@ -63,11 +65,17 @@ func (hc *httpClient) Do(req *http.Request, values url.Values) (*http.Response,
...
+	rdump, _ := httputil.DumpRequest(req, true)
+	fmt.Printf("%s\n", rdump)
+
 	resp, err := hc.client.Do(req)
 	if err != nil {
 		return nil, fmt.Errorf("client: %s resquest failed, %v", req.URL, err)
 	}
 
+	rsdump, _ := httputil.DumpResponse(resp, true)
+	fmt.Printf("%s\n", rsdump)
```
Comment (mtavano) [inline khipu.go]: this is a debug feature? ahahahaha

Comment (jaguirresza) [reply khipu.go]: yep

Comment (mtavano) [reply khipu.go]: nice 😂

**Comment by mtavano** [comment]
LRGTM 💃 💃 💃 💃 💃

**Pull request merged**